processing/src/main/java/org/apache/druid/query/groupby/ResultRowObjectMapperDecoratorUtil.java

```diff
@@ -0,0 +1,236 @@
+      // Have to deserialize map based rows, however don't have to deserialize the dimensions individually
+      // Deserializer that can deserialize either array- or map-based rows.
+      return new JsonDeserializer<ResultRow>()
+      {
+        @Override
+        public ResultRow deserialize(final JsonParser jp, final DeserializationContext ctxt) throws IOException
+        {
+          if (jp.isExpectedStartObjectToken()) {
+            final Row row = jp.readValueAs(Row.class);
+            return ResultRow.fromLegacyRow(row, query);
+          } else {
+            return ResultRow.of(jp.readValueAs(Object[].class));
+          }
+        }
+      };
+
+    } else {
+      // Have to deserialize the dimensions carefully
+      return new JsonDeserializer<ResultRow>()
+      {
+        final JavaType[] javaTypes = createJavaTypesForResultRow(query);
+
+        @Override
+        public ResultRow deserialize(final JsonParser jp, final DeserializationContext ctxt) throws IOException
+        {
+          if (jp.isExpectedStartObjectToken()) {
+            final Row row = jp.readValueAs(Row.class);
+            final ResultRow resultRow = ResultRow.fromLegacyRow(row, query);
+
+            final List<DimensionSpec> queryDimensions = query.getDimensions();
+            for (int i = 0; i < queryDimensions.size(); ++i) {
+              if (dimensionRequiresConversion(queryDimensions.get(i).getOutputType())) {
+                final int dimensionIndexInResultRow = query.getResultRowDimensionStart() + i;
+                resultRow.set(
+                    dimensionIndexInResultRow,
+                    objectMapper.convertValue(
+                        resultRow.get(dimensionIndexInResultRow),
+                        javaTypes[dimensionIndexInResultRow]
+                    )
+                );
+              }
+            }
+
+            return resultRow;
+          } else {
+            if (!jp.isExpectedStartArrayToken()) {
+              throw DruidException.defensive("Expected start token, received [%s]", jp.currentToken());
+            }
+
+            Object[] objectArray = new Object[query.getResultRowSizeWithPostAggregators()];
+            int index = 0;
+
+            while (jp.nextToken() != JsonToken.END_ARRAY) {
+              objectArray[index] = JacksonUtils.readObjectUsingDeserializationContext(jp, ctxt, javaTypes[index]);
+              ++index;
+            }
+
+            return ResultRow.of(objectArray);
+          }
+        }
+      };
+    }
+  }
+
+  @Nullable
+  private static JsonSerializer<ResultRow> getSerializer(
+      final GroupByQuery query,
+      final GroupByQueryConfig groupByQueryConfig
+  )
+  {
+    final boolean resultAsArray = query.context().getBoolean(GroupByQueryConfig.CTX_KEY_ARRAY_RESULT_ROWS, false);
+    final boolean intermediateCompatMode = groupByQueryConfig.isIntermediateResultAsMapCompat();
+    final boolean arrayBasedRows = resultAsArray && !intermediateCompatMode;
```

Comment:
not a blocker, but i wonder if we still need this config...

Reply:
We can probably remove the config safely and all the artifacts associated with maintaining the legacy version of the result rows.